qemu/tests/nic_teaming.py

```diff
@@ -0,0 +1,159 @@
+
+    1) Boot a vm with 4 nics.
+    2) inside guest, configure the team driver.
+    3) inside guest, ping host
+    4) inside guest, repeated down the slaves one by one.
+    5) check ping_result.
+
+    :param test: Kvm test object.
+    :param params: Dictionary with the test parameters.
+    :param env: Dictionary with test environment.
+    """
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    timeout = int(params.get("login_timeout", 1200))
+    session_serial = vm.wait_for_serial_login(timeout=timeout)
+    ifnames = [utils_net.get_linux_ifname(session_serial,
+                                          vm.get_mac_address(vlan))
+               for vlan, nic in enumerate(vm.virtnet)]
+    nm_stop_cmd = "pidof NetworkManager && service NetworkManager stop; true"
+    session_serial.cmd_output_safe(nm_stop_cmd)
+    team_if = params.get("team_if")
+    # initial
+
+    def team_port_add():
+        """Team0 add ports and return the ip link result for debuging"""
+        for port in ifnames:
+            session_serial.cmd_output_safe("ip addr flush %s" % port)
+            session_serial.cmd_output_safe("ip link set %s down" % port)
+            session_serial.cmd_output_safe("teamdctl %s port add %s"
+                                           % (team_if, port))
+        session_serial.cmd_output_safe("pkill dhclient")
+        session_serial.cmd_output_safe("dhclient -v %s" % team_if)
+        return session_serial.cmd_output_safe("ip addr ")
+
+    def failover(ifnames):
+        """func for failover"""
+        time.sleep(3)
+        while True:
+            pid_ping = session_serial.cmd_output_safe("pidof ping")
+            pid = re.findall(r"(\d+)", pid_ping)
+            try:
+                pid[0]
+            except IndexError:
+                break
+                # if ping finished, will break the loop.
+            for port in ifnames:
+                session_serial.cmd_output_safe("ip link set %s down" % port)
+                time.sleep(random.randint(5, 30))
+                session_serial.cmd_output_safe("ip link set %s up" % port)
+                time.sleep(1)
+
+    def check_ping(status, output):
+        """ ratio <5% is acceptance."""
+        if status != 0:
+            raise error.TestFail("Ping failed, staus:%s, output:%s"
+                                 % (status, output))
+        # if status != 0 the ping process seams hit issue.
+        ratio = utils_test.get_loss_ratio(output)
+        if ratio == -1:
+            raise error.TestFail('''The ratio is %s, and status is %s,
+                                    output is %s''' % (ratio, status, output))
+        elif ratio > 5:
+            raise error.TestFail("The loss raito is %s, test failed" % ratio)
+        logging.info("ping pass with loss raito:%s, that less than 5" % ratio)
+
+    def team_if_exist():
+        """ judge if team is alive well."""
+
+        pid_team = session_serial.cmd_output_safe("pidof teamd")
```

Comment:
in cfg:
team_if_exists_cmd =  ip link show ${team_if}
team_debug_cmd = "pidof teamd; teamdctl  ${team_if} config dump"
code:
team_exists_cmd = params.get("team_if_exists_cmd")
exits = session.cmd_status(team_exists_cmd) == 0 
if exits:
     output = session.cmd_output_safe(params["team_debug_cmd"])
     logging.debug("team interface configuration: %s" % output) 
     raise error.TestFail("team interface not exists, check debug info to check interface details.")